src/models/adVAE.jl

```diff
@@ -0,0 +1,187 @@
+            X = cat(X, rec_loss, dims=2)
+        end
+        output = cat(output, Flux.mean(X, dims=2), dims=1)
+    end
+    testmode!(advae, false)
+    return output
+end
+
+
+function StatsBase.fit!(advae::adVAE, data, params)
+	# prepare batches & loaders
+	train_loader, val_loader = prepare_dataloaders(data, params)
+	# training info logger
+	history = MVHistory()
+	# prepare for early stopping
+	best_adVAE = deepcopy(advae)
+	patience = params.patience
+	best_val_loss = 1e10
+	val_batches = length(val_loader)
+
+	# ADAMW(η = 0.001, β = (0.9, 0.999), decay = 0) = Optimiser(ADAM(η, β), WeightDecay(decay))
+	opt = haskey(params, :decay) ? ADAMW(params.lr, (0.9, 0.999), params.decay) : ADAM(params.lr)
+
+	ps_step1 = Flux.params(advae.generator, advae.transformer)
+	ps_step2 = Flux.params(advae.encoder)
+
+
+	progress = Progress(length(train_loader))
+	for (iter, X) in enumerate(train_loader)
+		# training step 1
+		loss_1, back = Flux.pullback(ps_step1) do
+            loss(advae|>gpu , getobs(X)|>gpu, γ=params.gamma, λ=params.lambda, mx=params.mx, mz=params.mz)
+		end
+		grad_step1 = back((1f0, 0f0, 0f0, 0f0))
+        Flux.Optimise.update!(opt, ps_step1, grad_step1)
+        # training step 2
+        loss_2, back = Flux.pullback(ps_step2) do
+            loss(advae|>gpu , getobs(X)|>gpu, γ=params.gamma, λ=params.lambda, mx=params.mx, mz=params.mz)
+		end
+        grad_step2 = back((0f0, 1f0, 0f0, 0f0))
+		Flux.Optimise.update!(opt, ps_step2, grad_step2)
+
+
+        push!(history, :loss_step_1, iter, loss_1[1])
+        push!(history, :encoder_loss, iter, loss_2[2])
+        push!(history, :generator_loss, iter, loss_1[3])
+        push!(history, :transformer_loss, iter, loss_1[4])
+
+		next!(progress; showvalues=[
+			(:iters, "$(iter)/$(params.iters)"),
+			(:loss_step_1, loss_1[1]),
+			(:loss_step_2, loss_2[2])
+			])
+        # TODO: check and prepare early stopping 
+        if mod(iter, params.check_every) == 0
+            total_val_loss = 0
+            testmode!(advae)
+            for X_val in val_loader
+                X_val = X_val |> gpu
+                μ, Σ = advae.encoder(X_val)
+                z = μ + Σ * randn(Float32)
+                xᵣ = advae.generator(z)
+                total_val_loss += Flux.mse(X_val, xᵣ)
+            end
+            testmode!(advae, false)
+            push!(history, :validation_loss, iter, total_val_loss/val_batches)
+
+			if total_val_loss < best_val_loss
+				best_val_loss = total_val_loss
+				patience = params.patience
+				best_adVAE = deepcopy(advae)
+			else
+				patience -= 1
+				if patience == 0
+					@info "Stopped training after $(iter) iterations"
+					global iters = iter - params.check_every*params.patience
+					break
+				end
+			end
+		end
+		if iter == params.iters
+			global iters = params.iters
+		end
+	end
+	return history, best_adVAE, sum(map(p->length(p), ps_step1))+sum(map(p->length(p), ps_step2)), iters
```

Comment:
If you have defined the model correctly with functors etc, you should only need to call `sum(map(p->length(p), Flux.params(best_adVAE)))`

Reply:
changed